sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -899,6 +919,21 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
           spec, partitionColumnNames, tablePath)
         try {
           tablePath.getFileSystem(hadoopConf).rename(wrongPath, rightPath)
+
+          // If the newSpec contains more than one depth partition, FileSystem.rename just deletes
+          // the leaf(i.e. wrongPath), we should check if wrongPath's parents need to be deleted.
+          // For example, give a newSpec 'A=1/B=2', after calling Hive's client.renamePartitions,
+          // the location path in FileSystem is changed to 'a=1/b=2', which is wrongPath, then
+          // although we renamed it to 'A=1/B=2', 'a=1/b=2' in FileSystem is deleted, but 'a=1'
+          // is still exists, which we also need to delete
+          val delHivePartPathAfterRename = getExtraPartPathCreatedByHive(
```

Comment:
`client.renamePartitions` is called at the beginning of `renamePartitions` for all specs at once. It creates the directory `a=1` and `a=1/b=2` and `a=1/b=3`.

When you iterates specs and rename the directories with FileSystem.rename, in the first iteration, `a=1/b=2` is renamed, and `a=1` is deleted in this change, then `a=1/b=3` will be deleted too. So in next iteration, the renaming of `a=1/b=3` to `A=1/B=3` will fail.